### PR TITLE
Load layer metadata information lazily 

### PR DESCRIPTION
**Pull request by iwehrman**
1. Load layer metadata lazily, so that when initializing a document metadata is only requested for selected layers. As with other lazy properties, metadata is loaded on-demand by `initializeLayers`/`resetLayers`. Addresses #3290. 
2. Respect the `lazy` parameter in `resetLayers` when choosing whether or not to load layer metadata when resetting layers.
3. Correctly reset/update layer export information as a result of `resetLayers`. Previously, updated export information that resulted from `resetLayers` was ignored.

This provides a modest document load-time speedup. For Vermilion, it seems to save a small few hundred milliseconds.

### CODE REVIEW

File: src/js/actions/layers.js [hunk 1]
```diff
@@ -206,50 +219,47 @@ define(function (require, exports) {
...
-        // And it is safe to ignore ALL bg layers because we don't use extension data on them
-        var nameSpace = global.EXTENSION_DATA_NAMESPACE,
-            numberOfLayers = doc.hasBackgroundLayer ? doc.numberOfLayers + 1 : doc.numberOfLayers,
-            indexRange = _.range(1, startIndex + numberOfLayers),
-            extensionPlayObjects = indexRange.map(function (i) {
-                var layerRef = layerLib.referenceBy.index(i);
-                return layerLib.getExtensionData(docRef, layerRef, nameSpace);
-            }),
+        var extensionPromise;
+        if (doc.hasBackgroundLayer && doc.numberOfLayers === 0 && targetLayers.length === 1) {
```
Comment (iwehrman) [inline src/js/actions/layers.js]: I special-cased this differently than you did previously, @mcilroyc. The reason is that it's fairly tricky to merge the lazy and metadata properties subset into the full properties set. I think this is actually a good, fast way to do it, but if something looks off to you please let me know.

File: src/js/actions/layers.js [hunk 2]
```diff
@@ -206,50 +219,47 @@ define(function (require, exports) {
...
-                return lazyPropertiesPromise.each(function (lazyProperties) {
-                    if (!lazyProperties) {
-                        // A background will not have a layer ID
-                        return;
-                    }
-                    
-                    var lazyLayerID = lazyProperties.layerID;
 
-                    _.merge(propertiesByID[lazyLayerID], lazyProperties);
+                return extensionPromise.then(function (allData) {
```
Comment (iwehrman) [inline src/js/actions/layers.js]: See "tricky merging" above.

File: src/js/models/documentexports.js [hunk 3]
```diff
@@ -67,6 +67,44 @@ define(function (require, exports, module) {
     }));
 
     /**
+     * Generate List of ExportAssets based on an array of simple props objects
```
Comment (iwehrman) [inline src/js/models/documentexports.js]: The following two functions were factored directly out of `fromDescriptors` below.

File: src/js/stores/export.js [hunk 4]
```diff
@@ -181,6 +183,23 @@ define(function (require, exports, module) {
                 documentExports = DocumentExports.fromDescriptors(payload);
 
             this._documentExportsMap = this._documentExportsMap.set(documentID, documentExports);
+            this.emit("change");
```
Comment (iwehrman) [inline src/js/stores/export.js]: This is an unrelated bugfix. @mcilroyc, you may wish to confirm that adding this change event doesn't have any ill effects.

File: src/js/actions/layers.js [hunk 5]
```diff
@@ -151,15 +159,20 @@ define(function (require, exports) {
             optionalPropertiesPromise = descriptor.batchMultiGetProperties(references, optionalProperties,
                 { continueOnError: true });
 
-        var nameSpace = global.EXTENSION_DATA_NAMESPACE,
-            extensionPlayObjects = references.map(function (ref) {
-                return layerLib.getExtensionData(ref[0], ref[1], nameSpace);
-            }),
+        var extensionPromise;
+        if (lazy) {
+            extensionPromise = Promise.resolve();
```
Comment (iwehrman) [inline src/js/actions/layers.js]: To test this case, note that all text layers (including non-selected ones) are reset when the font list is first initialized, and that happens when a text layer is first selected.

File: src/js/actions/layers.js [hunk 6]
```diff
@@ -206,50 +219,47 @@ define(function (require, exports) {
...
-                    return (extensionDataRoot && extensionDataRoot.exportsMetadata) || {};
-                })
-                .then(function (extensionDataArray) {
-                    // add an empty object associated with the background layer (see note above)
-                    if (startIndex === 0) {
-                        extensionDataArray.unshift({});
-                    }
-                    return extensionDataArray;
-                });
+                    .map(function (extensionData) {
```
Comment (mcilroyc) [inline src/js/actions/layers.js]: over-indented, methinks

File: src/js/actions/layers.js (hunk 2 above)
Comment (mcilroyc) [inline src/js/actions/layers.js]: gah

File: src/js/actions/layers.js (hunk 6 above)
Comment (iwehrman) [inline src/js/actions/layers.js]: You got me. Commit amended.

**Comment by mcilroyc** [comment]
looks good
